module purge; module load STAR/2.5.1b; module load java/latest; module load samtools/1.5; alias STAR='/home/tree101/scratch/STAR/STAR/bin/Linux_x86_64/STAR'; 
PATH=/home/tree101/scratch/STAR/STAR/bin/Linux_x86_64:$PATH;
#### for AS5 #### 
JOBID=$(sbatch --partition=nih_s10 --time=12:00:00 --workdir=/scratch/users/tree101/human/p3/trim/star_aligned/ --account=ascor --export=ALL -e /scratch/users/tree101/human/p3/trim/star_aligned/reports/s1_AS5.error -o /scratch/users/tree101/human/p3/trim/star_aligned/reports/s1_AS5_.o --job-name=st_AS5 --mem=71GB -n 8 --wrap 'STAR --genomeDir /srv/gsfs0/projects/sweet-cordero/alex/bin/index/STAR/hg38.p5/ --runThreadN 8 --readFilesIn /scratch/users/tree101/human/p3/trim/AS5_1.fq.gz /scratch/users/tree101/human/p3/trim/AS5_2.fq.gz --readFilesCommand zcat --outFileNamePrefix /scratch/users/tree101/human/p3/trim/star_aligned//AS5 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --sjdbGTFfile /srv/gsfs0/projects/sweet-cordero/alex/bin/index/STAR/hg38.p5/genecode_nopseudo.gtf' | awk '{print $4}')
JOBID=$(sbatch --partition=nih_s10 --time=12:00:00 --dependency=afterany:$JOBID --workdir=/scratch/users/tree101/human/p3/trim/star_aligned/ --account=ascor --export=ALL -e /scratch/users/tree101/human/p3/trim/star_aligned/reports/s1_AS5.error -o /scratch/users/tree101/human/p3/trim/star_aligned/reports/s1_AS5_.o --job-name=ngst_AS5 --mem=30GB -n 1 --wrap 'java -Xmx30g -jar /srv/gsfs0/projects/sweet-cordero/alex/bin/ngsutilsj bam-stats --silent --unique --gtf /srv/gsfs0/projects/sweet-cordero/alex/bin/index/STAR/hg38.p5/genecode_nopseudo.gtf -o /scratch/users/tree101/human/p3/trim/star_aligned/QC/AS5.stats.txt /scratch/users/tree101/human/p3/trim/star_aligned/AS5Aligned.sortedByCoord.out.bam' | awk '{print $4}');
#### for AS3 #### 
JOBID=$(sbatch --partition=nih_s10 --time=12:00:00 --workdir=/scratch/users/tree101/human/p3/trim/star_aligned/ --account=ascor --export=ALL -e /scratch/users/tree101/human/p3/trim/star_aligned/reports/s1_AS3.error -o /scratch/users/tree101/human/p3/trim/star_aligned/reports/s1_AS3_.o --job-name=st_AS3 --mem=71GB -n 8 --wrap 'STAR --genomeDir /srv/gsfs0/projects/sweet-cordero/alex/bin/index/STAR/hg38.p5/ --runThreadN 8 --readFilesIn /scratch/users/tree101/human/p3/trim/AS3_1.fq.gz /scratch/users/tree101/human/p3/trim/AS3_2.fq.gz --readFilesCommand zcat --outFileNamePrefix /scratch/users/tree101/human/p3/trim/star_aligned//AS3 --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --sjdbGTFfile /srv/gsfs0/projects/sweet-cordero/alex/bin/index/STAR/hg38.p5/genecode_nopseudo.gtf' | awk '{print $4}')
JOBID=$(sbatch --partition=nih_s10 --time=12:00:00 --dependency=afterany:$JOBID --workdir=/scratch/users/tree101/human/p3/trim/star_aligned/ --account=ascor --export=ALL -e /scratch/users/tree101/human/p3/trim/star_aligned/reports/s1_AS3.error -o /scratch/users/tree101/human/p3/trim/star_aligned/reports/s1_AS3_.o --job-name=ngst_AS3 --mem=30GB -n 1 --wrap 'java -Xmx30g -jar /srv/gsfs0/projects/sweet-cordero/alex/bin/ngsutilsj bam-stats --silent --unique --gtf /srv/gsfs0/projects/sweet-cordero/alex/bin/index/STAR/hg38.p5/genecode_nopseudo.gtf -o /scratch/users/tree101/human/p3/trim/star_aligned/QC/AS3.stats.txt /scratch/users/tree101/human/p3/trim/star_aligned/AS3Aligned.sortedByCoord.out.bam' | awk '{print $4}');
